### DISCUSSION TITLE
c# nullable reference types: narrowing of IEnumerable<T>? to IEnumerable<T>

### DISCUSSION THREAD

**Discussion by naighes** [dotnet/roslyn#68376] [General]
Howdy!
Not sure if this falls into the [known pitfalls of arrays](https://learn.microsoft.com/en-us/dotnet/csharp/nullable-references#arrays), but I cannot "unwrap" `IEnumerable<T>?` to `IEnumerable<T>`.
Here's what I'm doing:

```
IEnumerable<string>? x = new List<string>();
var y = x ?? Enumerable.Empty<string>();
```

Anyway, it looks `y` is still of type `IEnumerable<string>?` rather than `IEnumerable<string>`, despite the usage of the null-coalescing operator... and cannot sort it out why.
Any hint?

PS. by the way, without type inference the compiler doesn't complain anymore:

```
IEnumerable<string>? x = new List<string>();
IEnumerable<string> y = x ?? Enumerable.Empty<string>(); // no warnings
```

**Comment by svick** [comment]
When you use `var` and initialize the variable with a non-nullable reference type, it starts as "nullable, but known to be not null". This is so that you could assign `null` to it later on. But the compiler understands that it's not null, so it will mostly treat it as non-nullable otherwise.
